src/test/java/mate/academy/spring/boot/controller/ShoppingCartControllerTest.java

```diff
@@ -0,0 +1,309 @@
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.http.HttpStatus;
+import org.springframework.jdbc.datasource.init.ScriptUtils;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+import javax.sql.DataSource;
+import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class ShoppingCartControllerTest {
+    @Autowired
+    private ObjectMapper objectMapper;
+    private MockMvc mockMvc;
+
+    @BeforeEach
+    void beforeAll(
+            @Autowired DataSource dataSource,
+            @Autowired WebApplicationContext applicationContext
+    ) throws SQLException {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/add-users-to-users-table.sql"));
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/add-books-to-books-table.sql"));
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/add-shopping-carts-to-shopping-carts-table.sql"));
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/add-cart-items-to-cart-items-table.sql"));
+        }
+    }
+
+    @BeforeEach
+    void setup() {
+        User mockUser = new User();
+        mockUser.setId(3L);
+        mockUser.setEmail("user1@example.com");
+
+        SecurityContextHolder.getContext().setAuthentication(
+                new UsernamePasswordAuthenticationToken(mockUser, null, List.of())
+        );
+    }
```

Comment:
you can replace it with 
`@WithMockUser(username = "user1@example.com", roles = "USER")`

Reply:
If I run a test with this annotation, I get an error: jakarta.servlet.ServletException: Request processing failed: java.lang.NullPointerException: Cannot invoke "mate.academy.spring.boot.model.User.getId()" because "user" is null

Reply:
It should work correctly, so you can investigate this issue.